libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentBuilder.java

```diff
@@ -1302,4 +1303,45 @@ private static void ensureNoSelfReferences(final Iterable<?> value, Object origi
         }
     }
 
+    /**
+     * Checks whether the given value is writeable as x-content.
+     *
+     * If the value cannot be passed to {@link #value(Object)} without
+     * error, an error {@link IllegalArgumentException} is thrown.
+     */
+    public static void ensureToXContentable(@Nullable Object value) {
+        if (value == null) {
+            return;
+        }
+
+        if (value instanceof Object[] array) {
+            for (Object v : array) {
+                ensureToXContentable(v);
+            }
+        } else if (value instanceof Map<?, ?> map) {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                if (entry.getKey() instanceof String == false) {
+                    throw new IllegalArgumentException(
+                        "Cannot write non-String map key [" + entry.getKey().getClass().getCanonicalName() + "] to x-content"
```

Comment:
It's possible for a class to have a `null` canonical name (e.g. anonymous classes).
Should this just be `getName()` ?